components/gitpod-db/src/typeorm/user-db-impl.ts

```diff
@@ -350,6 +356,95 @@ export class TypeORMUserDBImpl implements UserDB {
     public async findUserByName(name: string): Promise<User | undefined> {
         return (await this.getUserRepo()).findOne({ name });
     }
+
+    // OAuthAuthCodeRepository
+    // OAuthUserRepository
+    public async getUserByCredentials(identifier: string, password?: string, grantType?: GrantIdentifier, client?: OAuthClient): Promise<OAuthUser | undefined> {
+        log.info(`getUserByCredentials ${identifier}`)
+        const user = await this.findUserById(identifier);
+        if (user) {
+            log.info(`getUserByCredentials returned: ${user.id}, ${user.name}`)
+            return {
+                id: user.id,
+                name: user.name,
+            };
+        }
+        log.info(`getUserByCredentials returned nothing!`)
+        return;
+    }
+    public async extraAccessTokenFields?(user: OAuthUser): Promise<ExtraAccessTokenFields | undefined> {
+        // No extra fields in token
+        log.info(`getUserByCredentials ${JSON.stringify(user)}`);
+        return;
+    }
+
+    // OAuthTokenRepository
+    async issueToken(client: OAuthClient, scopes: OAuthScope[], user?: OAuthUser): Promise<OAuthToken> {
```

Comment:
how does this relate to `AuthCodeRepositoryDB.issueToken`?

Reply:
I think you meant AuthCodeRepositoryDB.issueAuthCode?

The auth code is returned by the authorization process, which the client uses to request, along with the PKCE creds for validation, the actual API token - in our case the Gitpod token.